File: go/pkg/models/user_zone_reputation.go
package models

import (
	"math"
	"time"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

type UserZoneReputationName string

const (
	UserZoneReputationNameNeutral   UserZoneReputationName = "neutral"
	UserZoneReputationNameFriendly  UserZoneReputationName = "friendly"
	UserZoneReputationNameHonored   UserZoneReputationName = "honored"
	UserZoneReputationNameRevered   UserZoneReputationName = "revered"
	UserZoneReputationNameExalted   UserZoneReputationName = "exalted"
	UserZoneReputationNameLegendary UserZoneReputationName = "legendary"

	BaseReputationPoints   int = 100
	ReputationGrowthFactor int = 3
)

var UserZoneReputationNames = []UserZoneReputationName{
	UserZoneReputationNameNeutral,
	UserZoneReputationNameFriendly,
	UserZoneReputationNameHonored,
	UserZoneReputationNameRevered,
	UserZoneReputationNameExalted,
	UserZoneReputationNameLegendary,
}

type UserZoneReputation struct {
	ID                    uuid.UUID              `json:"id" gorm:"type:uuid;default:uuid_generate_v4()"`
	CreatedAt             time.Time              `json:"createdAt"`
	UpdatedAt             time.Time              `json:"updatedAt"`
	UserID                uuid.UUID              `json:"userId"`
	ZoneID                uuid.UUID              `json:"zoneId"`
	Level                 int                    `json:"level"`
	TotalReputation       int                    `json:"totalReputation"`
	ReputationOnLevel     int                    `json:"reputationOnLevel"`
	LevelsGained          int                    `json:"levelsGained" gorm:"-"`
	Name                  UserZoneReputationName `json:"name" gorm:"-"`
	ReputationToNextLevel int                    `json:"reputationToNextLevel" gorm:"-"`
}

func (u *UserZoneReputation) TableName() string {
	return "user_zone_reputations"
}

func (u *UserZoneReputation) AfterFind(tx *gorm.DB) (err error) {
	u.Name = u.GetReputationName()
	u.ReputationToNextLevel = u.GetReputationToNextLevel()
	return
}

func (u *UserZoneReputation) AddReputationPoints(points int) {
	u.TotalReputation += points
	u.ReputationOnLevel += points
	if u.ReputationOnLevel >= u.GetReputationToNextLevel() && u.Level < 6 {
		u.Level++
		u.LevelsGained++
		u.ReputationOnLevel = u.TotalReputation - u.GetReputationToNextLevel()
	}
}

func (u *UserZoneReputation) GetReputationToNextLevel() int {
	return BaseReputationPoints * int(math.Pow(float64(GrowthFactor), float64(u.Level)))
}

func (u *UserZoneReputation) GetReputationName() UserZoneReputationName {
	return UserZoneReputationNames[u.Level-1]
}

File: dart/sonar_web/lib/models/user_zone_reputation.dart
enum UserZoneReputationName {
  neutral,
  friendly,
  honored,
  revered,
  exalted,
  legendary;

  static UserZoneReputationName? fromString(String? value) {
    if (value == null) return null;
    return UserZoneReputationName.values.firstWhere(
      (e) => e.name == value,
      orElse: () => UserZoneReputationName.neutral,
    );
  }
}

class UserZoneReputation {
  final String id;
  final String createdAt;
  final String updatedAt;
  final String userId;
  final String zoneId;
  final int level;
  final int totalReputation;
  final int reputationOnLevel;
  final int levelsGained;
  final UserZoneReputationName name;
  final int reputationToNextLevel;

  const UserZoneReputation({
    required this.id,
    required this.createdAt,
    required this.updatedAt,
    required this.userId,
    required this.zoneId,
    required this.level,
    required this.totalReputation,
    required this.reputationOnLevel,
    required this.levelsGained,
    required this.name,
    required this.reputationToNextLevel,
  });

  factory UserZoneReputation.fromJson(Map<String, dynamic> json) {
    return UserZoneReputation(
      id: json['id'] as String,
      createdAt: json['createdAt'] as String,
      updatedAt: json['updatedAt'] as String,
      userId: json['userId'] as String,
      zoneId: json['zoneId'] as String,
      level: (json['level'] as num).toInt(),
      totalReputation: (json['totalReputation'] as num).toInt(),
      reputationOnLevel: (json['reputationOnLevel'] as num).toInt(),
      levelsGained: (json['levelsGained'] as num).toInt(),
      name: UserZoneReputationName.fromString(json['name'] as String?) ?? UserZoneReputationName.neutral,
      reputationToNextLevel: (json['reputationToNextLevel'] as num).toInt(),
    );
  }
}
